pkg/controller/ingress/ingress_controller_test.go

```diff
@@ -54,6 +63,104 @@ var (
+	}{
+		name: "Clean up old route and new route is generated",
+		state: []routev1.Route{
+			routev1.Route{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:   "to-be-reconciled-0",
+					Labels: map[string]string{networking.IngressLabelKey: name, serving.RouteLabelKey: name, serving.RouteNamespaceLabelKey: namespace},
+				},
+				Spec: routev1.RouteSpec{Host: domainName},
+			},
+			routev1.Route{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:   "no-remove-missing-label",
+					Labels: map[string]string{networking.IngressLabelKey: name, serving.RouteLabelKey: name},
+				},
+				Spec: routev1.RouteSpec{Host: "b.example.com"},
+			},
+			routev1.Route{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:   "no-remove-other-label",
+					Labels: map[string]string{networking.IngressLabelKey: "another", serving.RouteLabelKey: name, serving.RouteNamespaceLabelKey: namespace},
+				},
+				Spec: routev1.RouteSpec{Host: "c.example.com"},
+			},
+		},
+		want: []routev1.Route{
+			routev1.Route{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:   routeName0,
+					Labels: map[string]string{networking.IngressLabelKey: name, serving.RouteLabelKey: name, serving.RouteNamespaceLabelKey: namespace},
+				},
+				Spec: routev1.RouteSpec{Host: domainName},
+			},
+			routev1.Route{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:   "no-remove-missing-label",
+					Labels: map[string]string{networking.IngressLabelKey: name, serving.RouteLabelKey: name},
+				},
+				Spec: routev1.RouteSpec{Host: "b.example.com"},
+			},
+			routev1.Route{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:   "no-remove-other-label",
+					Labels: map[string]string{networking.IngressLabelKey: "another", serving.RouteLabelKey: name, serving.RouteNamespaceLabelKey: namespace},
+				},
+				Spec: routev1.RouteSpec{Host: "c.example.com"},
+			},
+		},
+	}
+
+	t.Run(test.name, func(t *testing.T) {
+		// Register operator types with the runtime scheme.
+		s := scheme.Scheme
+		s.AddKnownTypes(networkingv1alpha1.SchemeGroupVersion, defaultIngress)
+		s.AddKnownTypes(routev1.SchemeGroupVersion, &routev1.Route{})
+		s.AddKnownTypes(routev1.SchemeGroupVersion, &routev1.RouteList{})
+
+		// Create a fake client to mock API calls.
+		cl := fake.NewFakeClient(defaultIngress, &routev1.RouteList{Items: test.state})
+
+		// Create a Reconcile Ingress object with the scheme and fake client.
+		r := &ReconcileIngress{base: &common.BaseIngressReconciler{Client: cl}, client: cl, scheme: s}
+		// Mock request to simulate Reconcile() being called on an event for a
+		// watched resource .
+		req := reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      name,
+				Namespace: namespace,
+			},
+		}
+		if _, err := r.Reconcile(req); err != nil {
+			t.Fatalf("reconcile: (%v)", err)
+		}
+
+		routeList := &routev1.RouteList{}
+		err := cl.List(context.TODO(), &client.ListOptions{}, routeList)
+		assert.Nil(t, err)
+
+		routes := routeList.Items
+		assert.Equal(t, len(test.want), len(routes))
```

Comment:
Thanks, this is soooo much better 🙂 . I'd suggest to even use the `ElementsMatch` function to be able to remove all of the manual sorting and checking. Possible diff here: https://github.com/openshift-knative/knative-openshift-ingress/commit/c729300e613d20ed71b8599edee3f90c3c8dfdd1

Why?
- It makes the test a lot simpler and directly expresses the intent. 
- It also makes sure to check the entire objects, which is sometimes annoying to setup but almost always worth the hassle in terms of test coverage.
- Most importantly: It produces readable debug outcome. If this fails, it'll say something like "3 was not 2" which tells you nothing to debug. `ElementsMatch` will output a diff that you can use to see what's going one.